ambry-rest/src/main/java/com.github.ambry.rest/NettySslFactory.java

```diff
@@ -0,0 +1,198 @@
+import io.netty.handler.ssl.SslProvider;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+import java.security.KeyStore;
+import java.util.ArrayList;
+import java.util.Arrays;
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.SSLParameters;
+import javax.net.ssl.TrustManagerFactory;
+
+
+/**
+ * An implementation of {@link SSLFactory} that uses Netty's SSL libraries. This has the benefit of using OpenSSL
+ * instead of JDK's SSL implementation when the netty-tcnative library is loaded. OpenSSL shows
+ * significant performance enhancements over the JDK implementation.
+ */
+public class NettySslFactory implements SSLFactory {
+  private final SslContext nettyServerSslContext;
+  private final SslContext nettyClientSslContext;
+  private final SSLContext jdkSslContext;
+  private final String endpointIdentification;
+
+  /**
+   * Instantiate a {@link NettySslFactory} from a config.
+   * @param sslConfig the {@link SSLConfig} to use.
+   * @throws GeneralSecurityException
+   * @throws IOException
+   */
+  public NettySslFactory(SSLConfig sslConfig) throws GeneralSecurityException, IOException {
+    nettyServerSslContext = getServerSslContext(sslConfig);
+    nettyClientSslContext = getClientSslContext(sslConfig);
+    // Netty's OpenSsl based implementation does not use the JDK SSLContext so we have to fall back to the JDK based
+    // factory to support this method.
+    jdkSslContext = new JdkSslFactory(sslConfig).getSSLContext();
+
+    this.endpointIdentification =
+        sslConfig.sslEndpointIdentificationAlgorithm.isEmpty() ? null : sslConfig.sslEndpointIdentificationAlgorithm;
+  }
+
+  @Override
+  public SSLEngine createSSLEngine(String peerHost, int peerPort, Mode mode) {
+    SslContext context = mode == Mode.CLIENT ? nettyClientSslContext : nettyServerSslContext;
+    SSLEngine sslEngine = context.newEngine(ByteBufAllocator.DEFAULT, peerHost, peerPort);
+
+    if (mode == Mode.CLIENT) {
+      SSLParameters sslParams = sslEngine.getSSLParameters();
+      sslParams.setEndpointIdentificationAlgorithm(endpointIdentification);
+      sslEngine.setSSLParameters(sslParams);
+    }
+    return sslEngine;
+  }
+
+  @Override
+  public SSLContext getSSLContext() {
+    return jdkSslContext;
+  }
+
+  /**
+   * @param config the {@link SSLConfig}
+   * @return a configured {@link SslContext} object for a client.
+   * @throws GeneralSecurityException
+   * @throws IOException
+   */
+  private static SslContext getServerSslContext(SSLConfig config) throws GeneralSecurityException, IOException {
+    return SslContextBuilder.forServer(getKeyManagerFactory(config))
+        .trustManager(getTrustManagerFactory(config))
+        .sslProvider(SslProvider.JDK)
+        .ciphers(getCipherSuites(config))
+        .protocols(getEnabledProtocols(config))
+        .clientAuth(ClientAuth.REQUIRE)
```

Comment:
don't we have a config for this in `SSLConfig`?

Reply:
Seems like you have a function in this class (`getClientAuth()`) to handle this but it is unused?

Reply:
forgot to change this back after some debugging. fixed now